mavros/src/plugins/hil.cpp

```diff
@@ -0,0 +1,320 @@
+		hil_controls_msg->aux[2] = hil_controls.aux3;
+		hil_controls_msg->aux[3] = hil_controls.aux4;
+		hil_controls_msg->mode = hil_controls.mode;
+		hil_controls_msg->nav_mode = hil_controls.nav_mode;
+		// [[[end]]] (checksum: a2c87ee8f36e7a32b08be5e0fe665b5a)
+
+		hil_controls_pub.publish(hil_controls_msg);
+	}
+
+	void handle_hil_actuator_controls(const mavlink::mavlink_message_t *msg, mavlink::common::msg::HIL_ACTUATOR_CONTROLS &hil_actuator_controls) {
+		auto hil_actuator_controls_msg = boost::make_shared<mavros_msgs::HilActuatorControls>();
+
+		hil_actuator_controls_msg->header.stamp = m_uas->synchronise_stamp(hil_actuator_controls.time_usec);
+		for (int i = 0; i < 16; ++i) {
+			hil_actuator_controls_msg->controls[i] = hil_actuator_controls.controls[i];
+		}
+		hil_actuator_controls_msg->mode = hil_actuator_controls.mode;
+		hil_actuator_controls_msg->flags = hil_actuator_controls.flags;
+
+		hil_actuator_controls_pub.publish(hil_actuator_controls_msg);
+	}
+
+	//! Message specification: @p https://pixhawk.ethz.ch/mavlink/#HIL_GPS
+	void set_hil_gps(uint64_t time_boot_us, uint8_t fix_type,
+				int32_t lat, int32_t lon, int32_t alt,
+				uint16_t eph, uint16_t epv, uint16_t vel,
+				int16_t vn, int16_t ve, int16_t vd,
+				uint16_t cog,
+				uint8_t satellites_visible) {
+		mavlink::common::msg::HIL_GPS gps;
+
+		// there is no target sys in this mavlink message!
+
+		gps.time_usec = time_boot_us;
+		gps.fix_type = fix_type;
+		gps.lat = lat;
+		gps.lon = lon;
+		gps.alt = alt;
+		gps.eph = eph;
+		gps.epv = epv;
+		gps.vel = vel;
+		gps.vn = vn;
+		gps.ve = ve;
+		gps.vd = vd;
+		gps.cog = cog;
+		gps.satellites_visible = satellites_visible;
+
+		UAS_FCU(m_uas)->send_message_ignore_drop(gps);
+	}
+
+	/* -*- low-level send -*- */
+
+	//! Message specification: @p https://pixhawk.ethz.ch/mavlink/#HIL_STATE_QUATERNION
+	void set_hil_state_quaternion(uint64_t time_boot_us,
+				float qw, float qx, float qy, float qz,
+				float rollspeed, float pitchspeed, float yawspeed,
+				int32_t lat, int32_t lon, int32_t alt,
+				int16_t vx, int16_t vy, int16_t vz,
+				uint16_t ind_airspeed, uint16_t true_airspeed,
+				int16_t xacc, int16_t yacc, int16_t zacc) {
+		mavlink::common::msg::HIL_STATE_QUATERNION state_quat;
+
+		state_quat.time_usec = time_boot_us;
+
+		state_quat.attitude_quaternion[0] = qw;
+		state_quat.attitude_quaternion[1] = qx;
+		state_quat.attitude_quaternion[2] = qy;
+		state_quat.attitude_quaternion[3] = qz;
+
+		state_quat.rollspeed = rollspeed;
+		state_quat.pitchspeed = pitchspeed;
+		state_quat.yawspeed = yawspeed;
+
+		state_quat.lat = lat;
+		state_quat.lon = lon;
+		state_quat.alt = alt;
+
+		state_quat.vx = vx;
+		state_quat.vy = vy;
+		state_quat.vz = vz;
+
+		state_quat.ind_airspeed = ind_airspeed;
+		state_quat.true_airspeed = true_airspeed;
+
+		state_quat.xacc = xacc;
+		state_quat.yacc = yacc;
+		state_quat.zacc = zacc;
+
+		UAS_FCU(m_uas)->send_message_ignore_drop(state_quat);
+	}
```

Comment:
And that. It might be useful if it takes Eigen::Vector3d and Quaterniond at input and fills message.
But not this function. Just increase loc.